feat: Add utility methods for conversion to/from Prosemirror

tommoor opened the pull request:
As discussed here: https://discuss.yjs.dev/t/prosemirror-serializer/240

This PR adds new methods to aid getting content into and out of Y.Doc format for Prosemirror users. As part of this I had to move some more generic methods out of the `syncPlugin` and up to the `lib`, I also pulled the plugin keys to a higher level to avoid circular dependencies. 

`yDocToProsemirrorJSON` Is probably not _perfect_ yet, there might be an edge case I haven't covered, but that should become readily apparent in production usage soon.

Edit: I've published a fork with these utilities `@tommoor/y-prosemirror`

tommoor commented on src/interface.js:
```diff
@@ -0,0 +1,124 @@
+    return ydoc
+  }
+
+  return updateYFragment(type.doc, type, doc, new Map())
```
Perhaps the `mapping` argument for `updateYFragment` should be made optional?

tommoor commented on src/interface.js:
```diff
@@ -0,0 +1,124 @@
+    let response
+
+    // TODO: Must be a better way to detect text nodes than this
+    if (!item.nodeName) {
```
Open to suggestions here

tommoor commented on src/lib.js:
Note: `updateYFragment` now returns the `Y.Doc`, it didn't before

tommoor commented on src/lib.js:
These methods are all just moved from `syncPlugin` without edits, apart from `updateYFragment`, as noted below.

tommoor commented on src/plugins/undo-plugin.js:
```diff
@@ -31,10 +30,10 @@ export const yUndoPlugin = ({ protectedNodes = new Set(['paragraph']), trackedOr
-                              !(item.content.type instanceof Text ||
-                              (item.content.type instanceof XmlElement && protectedNodes.has(item.content.type.nodeName))) ||
-                              item.content.type._length === 0
+          !(item.content instanceof ContentType) ||
```
These formatting changes are the result of running `standard --fix`

tommoor commented:
I'd like to add some tests but the current test suite seems to be already failing on master…

dmonad commented:
Hi @tommoor 

thanks for adding this feature!

It takes some time for me to review a PR this huge. The refactor didn't make sense to me as I believe the ydoc⇔prosemirror binding internals shoulds live in `sync-plugin.js`. I'm very well aware that I have circular dependencies. Some bundlers throw warnings, but they are well supported by bundlers and they are actually supported by the spec itself. There are some caveats when working with circular dependencies, which is why you should avoid them if possible. But I'm not going to sacrifice a sound project structure to suppress the warning of circular dependencies. 

I reverted the refactoring & the formatting changes, and added some tests. The actual change is much simpler for me to comprehend: https://github.com/yjs/y-prosemirror/pull/29

This project is used by quite a lot of projects and I can't have it fail in production because I wasn't able to review the whole refactor. The next time maybe you could add fewer changes.

This is published in y-prosemirror@1.0.0 (it was time for a stable release)

tommoor commented:
> I reverted the refactoring & the formatting changes, and added some tests. The actual change is much simpler for me to comprehend

Yep, fair – just happy to have this merged. My logic at the time was that it felt strange to be using method from the y-sync plugin file inside of utils, and yes causes import loops – in hindsight it makes sense to reduce the diff.

dmonad commented:
No worries, you are not the first one who wanted to refactor one of my projects. I can't count how many PRs I rejected that added semicolons everywhere ^^

This is a helpful PR so thanks. Just saying this for the next one ;)
